### PR TITLE
Allow skeletons to be reused in frozen-partial scenarios (for inheritance margin)

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This PR now allows skeleton references to be reused in frozen-partial scenarios. Now, if we have a frozen partial project, and it has a p2p reference to a different language, we'll just reuse whatever skeleton we built for that language last, not caring if it is up to date or not with teh compilation for that project. This allows some amount of semantics to flow p2p, without taking on the full cost of producing the full skeleton.

The remote side of inheritance margin is also updated to use frozen-partial for the data it is operating on. This will hopefully get it to consume far less work producing skeletons in the p2p case.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs [hunk 1]
```diff
@@ -1002,16 +1002,24 @@ public CompilationInfo(Compilation compilation, bool hasSuccessfullyLoaded, Comp
...
+                    if (state.CompilationWithoutGeneratedDocuments is { } compilation)
+                        return compilation.ToMetadataReference(projectReference.Aliases, projectReference.EmbedInteropTypes);
+                }
+                else
+                {
+                    // Cross project reference.  We need a skeleton reference.  Skeletons are too expensive to
+                    // generate on demand.  So just try to see if we can grab the last generated skeleton for that
+                    // project.
+                    var properties = new MetadataReferenceProperties(aliases: projectReference.Aliases, embedInteropTypes: projectReference.EmbedInteropTypes);
+                    this.SkeletonReferenceCache.TryGetAlreadyBuiltMetadataReference(properties);
```
Comment (JoeRobich) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs:1022]: Should this be returned?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs:1022]: it should!

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs:1022]:
```suggestion
                    return this.SkeletonReferenceCache.TryGetAlreadyBuiltMetadataReference(properties);
```

File: src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs [hunk 2]
```diff
@@ -45,6 +45,14 @@ internal abstract partial class AbstractInheritanceMarginService
             ImmutableArray<(SymbolKey symbolKey, int lineNumber)> symbolKeyAndLineNumbers,
             CancellationToken cancellationToken)
         {
+            // If we're starting from a document, use it to go to a frozen partial version of it to lower the amount of
+            // work we need to do running source generators or producing skeleton references.
+            if (documentForGlobalImports != null)
+            {
+                documentForGlobalImports = documentForGlobalImports.WithFrozenPartialSemantics(cancellationToken);
```
Comment (sharwell) [inline src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs:52]: This is already a document with frozen partial semantics (per caller)

**Review by sharwell — DISMISSED** [review]

File: src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs:52]: note that for classification what we do is actually send a boolean over to OOP saying taht it should be frozen-partial again on its end.

this is needed as frozen partial effectively means 'capture whatever completed computations are available'.  but hte set of completed computations is different on host/oop.  so if OOP hasn't performed some comptuations we don't want it to now pay that price to try to get it's computations in sync with the host.

Comment (sharwell) [reply src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs:52]: > note that for classification what we do is actually send a boolean over to OOP saying taht it should be frozen-partial again on its end.

We do this for several features, and should match it here.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs:52]: will do.

**Comment by CyrusNajmabadi** [comment]
@sharwell  for another look.

Note: as part of this PR i was finding the dataflow over different operations confusing.  So i split this up into two remote calls.  One for getting global imports, the other for getting symbol information.

**Comment by CyrusNajmabadi** [comment]
@Cosifne @sharwell this is ready for review.

File: src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs [hunk 3]
```diff
@@ -38,26 +38,65 @@ internal abstract partial class AbstractInheritanceMarginService
...
             TextSpan spanToSearch,
+            bool frozenPartialSemantics,
+            CancellationToken cancellationToken)
+        {
+            var solution = document.Project.Solution;
+            var remoteClient = await RemoteHostClient.TryGetClientAsync(solution.Workspace.Services, cancellationToken).ConfigureAwait(false);
+            if (remoteClient != null)
+            {
+                // Also, make it clear to the remote side that they should be using frozen semantics, just like we are.
+                // we want results quickly, without waiting for the entire SG pass to go.  The user will still get
```
Comment (Cosifne) [inline src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs:52]: `entire SG`
Sorry, could you explain what's 'SG'?

Comment (Cosifne) [reply src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs:52]: I think I know what it is after reading the following code, is it source generator? Could you inline the full name in the comments?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs:52]: Yup.  It's SourceGenerator.  i can definitely expand that in the comment :)

Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs:52]:
```suggestion
                // we want results quickly, without waiting for the entire source generator pass to run.  The user will still get
```

File: src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs [hunk 4]
```diff
@@ -38,26 +38,65 @@ internal abstract partial class AbstractInheritanceMarginService
...
-                // To match the returned inheritance targets to the line number, we need set an 'Id' when calling the remote process,
-                // however, given the line number is just an int, setting up an int 'Id' for an int is quite useless, so just passed it to the remote process.
+                // Here the line number is also passed to the remote process. It is done in this way because when a set
+                // of symbols is passed to remote process, those without inheritance targets would not be returned. To
+                // match the returned inheritance targets to the line number, we need set an 'Id' when calling the
+                // remote process, however, given the line number is just an int, setting up an int 'Id' for an int is
+                // quite useless, so just passed it to the remote process.
+                //
+                // Also, make it clear to the remote side that they should be using frozen semantics, just like we are.
+                // we want results quickly, without waiting for the entire SG pass to go.  The user will still get
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs:93]:
```suggestion
                // we want results quickly, without waiting for the entire source generator pass to run.  The user will still get
```

File: src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs]: done.

File: src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs [hunk 5]
```diff
@@ -69,28 +108,30 @@ internal abstract partial class AbstractInheritanceMarginService
...
-            Document? documentForGlobalImports,
-            TextSpan spanToSearch,
+            Document? document,
             ImmutableArray<(SymbolKey symbolKey, int lineNumber)> symbolKeyAndLineNumbers,
+            bool frozenPartialSemantics,
             CancellationToken cancellationToken)
         {
+            // If we're starting from a document, use it to go to a frozen partial version of it to lower the amount of
+            // work we need to do running source generators or producing skeleton references.
+            if (document != null && frozenPartialSemantics)
```
Comment (Cosifne) [inline src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs:125]: I feel the 'frozenPartialSemantics' pattern is somewhat strange to me. (The flag seems always to be true)
Is this the pattern that Sam mentioned in the previous comment?

**Review by Cosifne — APPROVED** [review]
LGTM except for the strange always true flag.
But if this is our pattern then that's fine : )

**Pull request merged**